### DISCUSSION TITLE
Babashka + org-mode literate programming example

### DISCUSSION THREAD

**Discussion by adam-james-v** [babashka/babashka#907] [Show and tell]
Hello,

The other day I made a file ob-babashka.el after someone on the Clojurian's Slack asked about babashaka code blocks in org-mode.

Using this I then wrote up a little example. In it I draw a flower. It's a simple project that has no particular goal, but I think it can be a useful starting point for anyone who may be interested in writing some clojure programs in a literate style.

ob-babashka.el:
https://gist.github.com/adam-james-v/f4d2b75a70b095d14a351a1eff96b4b0

bb-draw.org:
https://github.com/adam-james-v/bb-draw/blob/main/bb-draw.org

**Comment by mmzsource** [comment]
This is my config for literate programming clojure in .org files:

```
#+BEGIN_SRC emacs-lisp
(org-babel-do-load-languages
  'org-babel-load-languages
    '((emacs-lisp . t)
      (shell . t)
      (clojure . t)))
#+END_SRC

Don't ask before evaluating code blocks:

#+BEGIN_SRC emacs-lisp
(setq org-confirm-babel-evaluate nil)
#+END_SRC

#+BEGIN_SRC emacs-lisp
(message "%s" "Start loading ob-clojure")
(use-package ob-clojure
  :init
  (setq org-babel-clojure-backend 'cider))
#+END_SRC
```
First start a `bb --nrepl-server`  on the cmdline, then connect with cider-connect and now you can evaluate the `#+BEGIN_SRC clojure`  clojure blocks in your .org file. The nrepl-server will retain state over different code blocks.

**Reply by jakub-stastny** [reply]
Well that one is pretty obvious. However the purpose of this is to be able to run babashka scripts with `C-c C-c` and get results back.

My primary reason for wanting this was to simply run quick-and-dirty babashka scripts this way, without having to start REPL, since I use babashka scripts in many different parts of my system now.

The real strength of this is that it gives you results backs, which you can mix with other code blocks doing processing on them, whatever the language – so for instance you can get some data from babashka, postprocess it via a quick-and-dirty ZSH script and then render it as a graph in R.

REPL's great, but the use case of this is quite different.

**Reply by adam-james-v** [reply]
Thanks for sharing this approach, @mmzsource 

It may be obvious to people who know org-mode fairly well already, but it's nice to have snippets showing alternative approaches :)

I *hope* that this example stands out as just one of several equally valid ways of doing literate programs with Clojure.

The flexibility of emacs/ org-mode, CLJ,CLJS,babashka were a bit overwhelming for me when I was first diving into Clojure, so a simple(ish) example seems useful :)

Either way, we should always try to find the right tools for the job.

**Comment by mpas** [comment]
Nice work, currently i am using the following code block using a shell.

```
#+begin_src shell :shebang "#!/usr/bin/env bb"
(+ 1 2 3 4)
#+end_src
```

![image](https://user-images.githubusercontent.com/139094/124100555-dcbbac80-da5e-11eb-9f94-3ccae50bab7a.png)

**Reply by jakub-stastny** [reply]
@mpas that's an interesting workaround.

That wouldn't apply the correct major mode for the block though, or does it? Such as you don't get syntax highlighting and anything that you might be using for CJ editing (paredit/lispy, rainbow brackets etc).

**Reply by mmzsource** [reply]
@jakub-stastny In my case I do get some paren help in the shell block (possibly because of my paredit config?) but no syntax highlighting indeed. In my case that's not a real problem; the code blocks I'm writing / using are small and executing them in the org document is mostly a convenience feature. 
It does make me wonder: what are your use cases? Maybe I'm missing out here ... For a script or a project I move to .clj files. How big are your "quick-and-dirty babashka scripts"? Do you 'tangle' them? Why would you do this 'piping' between babashka, zsh & R scripts in an org file and not on the command line? What's your workflow? Is this maybe some reproducible science workflow? I like org mode a lot, and it seems you move it to next levels so I'm curious to know.

**Comment by cfclrk** [comment]
There was a patch submitted to the org-mode mailing list to add some (really cool) support for babashka code blocks, as well as code blocks using `clj` and deps:

- https://list.orgmode.org/87imctk1jb.fsf@gnu.org/T/

I'm using a slightly modified version of that patch myself, and it's been great so far.
